### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by bonky42** [F5Networks/k8s-bigip-ctlr#551]
#### Description
We are seeing an increase in memory usage, it go from 100 MiB to 2.5 GiB in 5 hours and we have only 4 ingress configuration, is their is a missing parameter on our side, we haven't put memory request limit yet? 

#### Kubernetes Version
1.7

#### Controller Version
1.3 (we cannot go to 1.4, it seems there is an incompability when updating policies on our bigip version)

#### BIG-IP Version
BIG-IP 11.6.1 Build 2.0.338 Hotfix HF2

#### Diagnostic Information
![image](https://user-images.githubusercontent.com/4395018/35965672-ea7de708-0cbb-11e8-995d-d2c765667c4e.png)

**Comment by f5yacobucci** [comment]
Can you provide us with a redacted Deployment/Pod configuration?

Also, can you tell us more about the policy incompatibility?

**Comment by f5yacobucci** [comment]
Do you see this behavior with a fairly static environment; 4 Ingress that don't often change and Services/Pods that don't often scale? Or is the environment very dynamic with Services/Pods that are scaling often and Ingress that are created and deleted?

**Comment by russokj** [comment]
Here's the controller object allocation shortly after starting:
```
(pprof) top 20
316957 of 322709 total (98.22%)
Dropped 57 nodes (cum <= 1613)
Showing top 20 nodes out of 98 (cum >= 32768)
      flat  flat%   sum%        cum   cum%
     98304 30.46% 30.46%     100352 31.10%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.(*jsonDecDriver).DecodeBytes
     80284 24.88% 55.34%      80284 24.88%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.(*jsonDecDriver).DecodeString
     32768 10.15% 65.49%      32768 10.15%  encoding/json.stateBeginValue
     32768 10.15% 75.65%      32768 10.15%  strings.genSplit
     16384  5.08% 80.73%      62578 19.39%  encoding/json.(*decodeState).object
     13653  4.23% 84.96%      13653  4.23%  time.NewTimer
     11606  3.60% 88.55%      11606  3.60%  encoding/json.(*decodeState).objectInterface
      8192  2.54% 91.09%      33862 10.49%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.fastpathT.DecSliceStringV
      6555  2.03% 93.12%      69133 21.42%  encoding/json.Unmarshal
      4096  1.27% 94.39%       4096  1.27%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest.(*Request).URL
      3641  1.13% 95.52%       3641  1.13%  github.com/F5Networks/k8s-bigip-ctlr/vendor/golang.org/x/net/http2.(*clientConnReadLoop).handleResponse
      2341  0.73% 96.25%      50402 15.62%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/pkg/api/v1.codecSelfer1234.decSliceNodeCondition
      2048  0.63% 96.88%       2048  0.63%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.(*jsonDecDriver).appendStringAsBytes
      1638  0.51% 97.39%       1638  0.51%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/pkg/api/v1.codecSelfer1234.decResourceList
      1024  0.32% 97.70%      34886 10.81%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/pkg/api/v1.codecSelfer1234.decSliceContainerImage
       910  0.28% 97.99%      62578 19.39%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/go-openapi/spec.(*Schema).UnmarshalJSON
       745  0.23% 98.22%     171776 53.23%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/pkg/api/v1.codecSelfer1234.decSliceNode
         0     0% 98.22%      32768 10.15%  crypto/x509.ParseCertificate
         0     0% 98.22%      32768 10.15%  crypto/x509.parseCertificate
         0     0% 98.22%      32768 10.15%  encoding/asn1.Unmarshal
```
10 minutes later, the view is the following:
```
(pprof) top 20
982407 of 1378057 total (71.29%)
Dropped 127 nodes (cum <= 6890)
Showing top 20 nodes out of 240 (cum >= 96442)
      flat  flat%   sum%        cum   cum%
    229378 16.65% 16.65%     231426 16.79%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.(*jsonDecDriver).DecodeBytes
    191463 13.89% 30.54%     199655 14.49%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.(*jsonDecDriver).DecodeString
     65536  4.76% 35.29%      82928  6.02%  net/http.(*Client).doFollowingRedirects
     55981  4.06% 39.36%     101622  7.37%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest.(*Request).URL
     37449  2.72% 42.07%      37449  2.72%  regexp/syntax.(*parser).collapse
     35498  2.58% 44.65%      35498  2.58%  github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager.createRule
     32768  2.38% 47.03%     136305  9.89%  encoding/json.(*decodeState).object
     32768  2.38% 49.41%      32768  2.38%  encoding/json.stateBeginValue
     32768  2.38% 51.78%      32768  2.38%  encoding/json.unquote
     32768  2.38% 54.16%      41508  3.01%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/apis/meta/v1.(*Time).UnmarshalJSON
     32768  2.38% 56.54%      65536  4.76%  regexp.mergeRuneSets
     32768  2.38% 58.92%      32768  2.38%  regexp.mergeRuneSets.func2
     32768  2.38% 61.30%      32768  2.38%  strings.genSplit
     24576  1.78% 63.08%      24576  1.78%  strings.Join
     21250  1.54% 64.62%      21250  1.54%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/transport.cloneRequest
     19505  1.42% 66.04%      19505  1.42%  net/url.(*URL).String
     19114  1.39% 67.42%      19114  1.39%  time.NewTimer
     17927  1.30% 68.72%      17927  1.30%  bytes.makeSlice
     17874  1.30% 70.02%      34117  2.48%  encoding/json.Marshal
     17480  1.27% 71.29%      96442  7.00%  encoding/json.Unmarshal
```
Still trying to get some meaning from this.

**Comment by bonky42** [comment]
> Can you provide us with a redacted Deployment/Pod configuration?

`---
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: k8s-bigip-ctlr-deployment
  labels:
    app: k8s-bigip-ctlr
  namespace: kube-system
spec:
  replicas: 1
  template:
    metadata:
      name: k8s-bigip-ctlr
      labels:
        app: k8s-bigip-ctlr
    spec:
      serviceAccountName: bigip-ctlr-serviceaccount
      containers:
        - name: k8s-bigip-ctlr
          # replace the version as needed
          image: "f5networks/k8s-bigip-ctlr:1.3.0"
          env:
            - name: BIGIP_USERNAME
              valueFrom:
                secretKeyRef:
                  name: bigip-login
                  key: username
            - name: BIGIP_PASSWORD
              valueFrom:
                secretKeyRef:
                  name: bigip-login
                  key: password
          command: ["/app/bin/k8s-bigip-ctlr"]
          args:
            - "--bigip-username=$(BIGIP_USERNAME)"
            - "--bigip-password=$(BIGIP_PASSWORD)"
            - "--bigip-url=********"
            - "--bigip-partition=********"
            - "--log-level=ERROR"
            #- "--default-ingress-ip=************"
            # To manage a single namespace, enter it below
            # (required in v1.0.0)
            # To manage all namespaces, omit the `namespace` entry
            # (default as of v1.1.0)
            # To manage multiple namespaces, enter a separate flag for each
            # namespace below (as of v1.1.0)
            #- "--namespace=default"
`

> Also, can you tell us more about the policy incompatibility?

Cannot reproduce it, good news, will give it a try, maybe there is not the same behavior for memory usage

> Do you see this behavior with a fairly static environment; 4 Ingress that don't often change and Services/Pods that don't often scale? Or is the environment very dynamic with Services/Pods that are scaling often and Ingress that are created and deleted?

It's fairly static, we put them and don't touch them anymore.

**Comment by f5yacobucci** [comment]
We'd like to see a more precise state of the container and see which portion of the controller is out of control - we have been unable to reproduce anything quite so dramatic as your graphs.

Find the node running the container:
`kubectl -n kube-system get pods; kubectl -n kube-system get pod [POD_NAME]`
The `status.hostIP` will show the host node IP.

SSH to node.

Find the container ID:
`sudo docker ps | grep "/app/bin/k8s-bigip"`

Exec into container:
`sudo docker exec -it -u 0 [CONTAINER_ID] sh`

Run top:
/app# top

What are the VSZ, %VSZ, CPU, and %CPU of the k8s-bigip-ctlr and python bigipconfigdriver processes?

Thanks for the help.

**Comment by russokj** [comment]
Here's the current allocated memory at the start ('top' indicates about 160MB dirty for ctlr, 80 MB for python driver):
```
bug/pprof/heapol pprof -inuse_objects /build/out/linux-amd64-release-go1.7.5/bin/k8s-bigip-ctlr http://10.190.24.163:31732/de
Fetching profile from http://10.190.24.163:31732/debug/pprof/heap
Saved profile in /home/user/pprof/pprof.k8s-bigip-ctlr.10.190.24.163:31732.inuse_objects.inuse_space.002.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top 20
100557 of 100627 total (99.93%)
Dropped 419 nodes (cum <= 503)
Showing top 20 nodes out of 56 (cum >= 43753)
      flat  flat%   sum%        cum   cum%
     38230 37.99% 37.99%      40680 40.43%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartRecordingToSink
     12290 12.21% 50.21%      22832 22.69%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*eventLogger).eventObserve
     10923 10.85% 61.06%      10923 10.85%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/sets.NewString
     10084 10.02% 71.08%      10084 10.02%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/golang/groupcache/lru.(*Cache).Add
      3782  3.76% 74.84%       3782  3.76%  runtime.malg
      3641  3.62% 78.46%       3641  3.62%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/sets.String.Insert
      3073  3.05% 81.51%       4562  4.53%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*events).PatchWithEventNamespace
      3073  3.05% 84.57%       3073  3.05%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*recorderImpl).makeEvent
      3073  3.05% 87.62%      50738 50.42%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.recordToSink
      2979  2.96% 90.58%       2979  2.96%  encoding/json.Marshal
      2450  2.43% 93.01%       2450  2.43%  math/rand.NewSource
      2284  2.27% 95.28%       2284  2.27%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/watch.(*Broadcaster).Watch.func1
      1821  1.81% 97.09%       1821  1.81%  bytes.makeSlice
      1489  1.48% 98.57%       1489  1.48%  net/textproto.MIMEHeader.Set
      1365  1.36% 99.93%       2854  2.84%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest.NewRequest
         0     0% 99.93%       1821  1.81%  bytes.(*Buffer).Write
         0     0% 99.93%       1821  1.81%  bytes.(*Buffer).grow
         0     0% 99.93%       1821  1.81%  encoding/json.(*Encoder).Encode
         0     0% 99.93%      43753 43.48%  github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager.(*Manager).(github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager.virtualServerWorker)-fm
         0     0% 99.93%      43753 43.48%  github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager.(*Manager).processNextVirtualServer
```
After an overnight run (continuous updating ingress and the number of service instances), the dirty memory reported by top showed 3027 MB for ctlr, 796 MB for python driver, with the allocated memory broken down as follows:
```
ebug/pprof/heapol pprof -inuse_objects /build/out/linux-amd64-release-go1.7.5/bin/k8s-bigip-ctlr http://10.190.24.163:31732/de
Fetching profile from http://10.190.24.163:31732/debug/pprof/heap
Saved profile in /home/user/pprof/pprof.k8s-bigip-ctlr.10.190.24.163:31732.inuse_objects.inuse_space.007.pb.gz
Entering interactive mode (type "help" for commands)
(pprof) top 40
3690240 of 3722364 total (99.14%)
Dropped 472 nodes (cum <= 18611)
Showing top 40 nodes out of 87 (cum >= 365916)
      flat  flat%   sum%        cum   cum%
    557076 14.97% 14.97%     612165 16.45%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*eventBroadcasterImpl).StartRecordingToSink
    365916  9.83% 24.80%     365916  9.83%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/ugorji/go/codec.(*jsonDecDriver).DecodeString
    282330  7.58% 32.38%     544485 14.63%  github.com/F5Networks/k8s-bigip-ctlr/vendor/github.com/golang/groupcache/lru.(*Cache).Add
    262155  7.04% 39.42%     262155  7.04%  container/list.(*List).PushFront
    257922  6.93% 46.35%     257922  6.93%  strings.Join
    225308  6.05% 52.41%     673529 18.09%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*eventLogger).eventObserve
    189180  5.08% 57.49%     189180  5.08%  net/textproto.MIMEHeader.Set
    186523  5.01% 62.50%     186523  5.01%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/watch.(*Broadcaster).Watch.func1
    180228  4.84% 67.34%     850540 22.85%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*EventAggregator).EventAggregate
    142000  3.81% 71.15%     142000  3.81%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/sets.NewString
    105600  2.84% 73.99%     105600  2.84%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/apimachinery/pkg/util/sets.String.Insert
    101054  2.71% 76.71%     279313  7.50%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest.NewRequest
     75370  2.02% 78.73%      75370  2.02%  fmt.Sprintf
     68526  1.84% 80.57%      68526  1.84%  encoding/json.Marshal
     65562  1.76% 82.33%      65562  1.76%  runtime.malg
     65537  1.76% 84.09%     801635 21.54%  github.com/F5Networks/k8s-bigip-ctlr/pkg/appmanager.(*Manager).recordIngressEvent
     58396  1.57% 85.66%      74780  2.01%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*recorderImpl).makeEvent
     58396  1.57% 87.23%    2476262 66.52%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.recordToSink
     55089  1.48% 88.71%      55089  1.48%  math/rand.NewSource
     54615  1.47% 90.18%     116200  3.12%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest.(*Request).Body
     54615  1.47% 91.65%     243795  6.55%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/rest.(*Request).SetHeader
     54615  1.47% 93.11%    1578684 42.41%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/tools/record.(*EventCorrelator).EventCorrelate
     49154  1.32% 94.43%      49154  1.32%  time.Sleep
     49153  1.32% 95.75%      49153  1.32%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*CoreV1Client).Events
     45878  1.23% 96.99%      45878  1.23%  runtime.deferproc.func1
     28817  0.77% 97.76%      28817  0.77%  bytes.makeSlice
     28686  0.77% 98.53%     381733 10.26%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*events).CreateWithEventNamespace
     22539  0.61% 99.14%     441064 11.85%  github.com/F5Networks/k8s-bigip-ctlr/vendor/k8s.io/client-go/kubernetes/typed/core/v1.(*events).PatchWithEventNamespace
         0     0% 99.14%      28817  0.77%  bytes.(*Buffer).Write
```

**Comment by bonky42** [comment]
Sorry for the delay, I can't see for now (infrastructure is migrating, I don't have access to kubernetes for some times). Just one thing I remember, we've got an incident on the F5 around the same time admin interface was not working anymore and so the webservices (we've got a case open for this one). Maybe the memory increase if the controller can't join the F5.

**Comment by russokj** [comment]
Thanks for the update.  We will take a look at the F5 connection to see if it plays a part in the memory consumption. As it is, we have verified a few issues with in regards to memory leakage with 1.3 of the controller.  The latest version (1.4) handles memory much better in the controller, although we are still tracking down an issue with the python driver.

**Comment by f5yacobucci** [comment]
Update: yes, we have fixed controller memory management in the k8s-bigip-ctlr 1.4 main process. The subprocess we use to interface with our REST client was showing memory growth in 1.4. We identified an updated sub-dependency that was at fault. We expect a fix today, will be upgrading the dependency and conducting tests before releasing a new patch - timing is dependent on the fix that should land today.

**Comment by russokj** [comment]
@bonky42 - We just released version 1.4.2 that contains fixes for the memory issues you encountered (or would have) with versions 1.3.0 - 1.4.1 of the controller.

**Comment by bonky42** [comment]
Ok, thanks, I will give it a try as soon as possible.

Le 27 févr. 2018 10:57 AM, "Ken Russo" <notifications@github.com> a écrit :

@bonky42 <https://github.com/bonky42> - We just released version 1.4.2 that
contains fixes for the memory issues you encountered (or would have) with
versions 1.3.0 - 1.4.1 of the controller.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/F5Networks/k8s-bigip-ctlr/issues/551#issuecomment-368987262>,
or mute the thread
